### ISSUE TITLE
Implement SCSS lint

### ISSUE THREAD

**Issue by akibrhast** [hackforla/website#1441]
### Overview
We need to implement a linter for our SCSS code such that code is more maintainable and standardized over time.

### Action Items
- Create a list of possible bare minimum SCSS linter settings that we can turn on, get an output of running those SCSS linter setting on our current codebase and comment on this issue the esLint settings and the output from running it.
- Implement a github action that runs the SCSS  linter on the files changed of the pull request if that file has a `.scss` extension

### Resources/Instructions
 <sub>As you are working on this issue, update this section with the resources and instructions that you come across and use when to implement this functionality. </sub>
CSS Linter (stylelint): https://stylelint.io/
GitHub action (Super-Linter): https://github.com/marketplace/actions/super-linter
How to Setup Article with Video: https://dev.to/n3wt0n/the-easiest-way-to-lint-any-code-github-super-linter-deep-dive-53eo

**Comment by averdin2** [comment]
So far, I am looking at using this GitHub [Action](https://github.com/marketplace/actions/lint-action#limitations). There seems to be a known issue with using Forks with this GitHub Action, so I need to look into it a little more.

As far as the linter itself, [Stylelint](https://stylelint.io/) seems like the way to go.

As far as the rules we should implement, no-duplicate-selectors, selector-pseudo-class-no-unknown, selector-pseudo-element-no-unknown, and selector-type-no-unknown seem to be the minimum rules we are looking for. Here is a [link](https://stylelint.io/user-guide/rules/list) for the list of Stylelint rules. You can find the ones I mentioned in that link.

I still need to determine if these rules are adequate, and how to implement them with GitHub Actions.

**Comment by averdin2** [comment]
ETA: I probably won't have another update for this until Thursday. 
Progress Plan: I am planning on implementing this GitHub Action in my own repo, and I also want to attend the GitHub actions talk during the Engineering COP meeting on Wednesday to get some more knowledge on using GitHub actions.

**Comment by averdin2** [comment]
I set up a GitHub action on my own [repo](https://github.com/averdin2/Web_Dev_Practice). For the actual GitHub action, I went with [Super-Linter](https://github.com/marketplace/actions/super-linter). In my initial research, I didn't find this GitHub action, but I think it is way better than the previous one I found as it was made by GitHub. I am also still sticking with [Stylelint](https://stylelint.io/) as this is the leading linter for CSS.

I was able to get the GitHub action to trigger for good linting and bad linting for pull requests given our specific needs. One complication I have run into is that Stylelint does not have a default rule for checking unused CSS classes (one of the major points of why we want to lint). I found a potential [plugin](https://www.npmjs.com/package/stylelint-no-unused-selectors) that I could use, but I need to do some more research on this topic.

Below I have added a screenshot of my current YMAL file to configure my current iteration of the linting. I made some notes on the image where changes for the final version would need to be made.

<details>
<summary>Image of linter.yml</summary>

![image](https://user-images.githubusercontent.com/10714770/117878158-49ce7500-b25a-11eb-8715-8df397a35d42.png)
</details>

**Comment by averdin2** [comment]
I am making a comment for some action items I still need to do.
#### Action Items
- [x] Set up SCSS in my personal repo and test linter
- [x] Test linting GitHub action check with forked PRs
- [ ] Figure out any other potential CSS linting rules wanted
- [x] See if plugin will work for linting unused CSS classes
- [ ] Research how to use above plugin with GitHub Actions
- [x] Research other ways to lint unused CSS classes
- [ ] Implement changes into website repo

**Comment by averdin2** [comment]
@sanchece here is that other ES Lint issue if you wanted to look into this a bit (https://github.com/hackforla/website/issues/1442). Again, I was thinking of implementing this with Super-Linter. I think it should actually not take too long once I figure out if the CSS linter will work.

**Comment by qiqicodes** [comment]
@averdin2 Can you let us know your progress in regards to this issue? Thank you =]

Progress:
Blockers:

**Comment by averdin2** [comment]
The Stylelint [plugin](https://www.npmjs.com/package/stylelint-no-unused-selectors) I was looking into seems to not work for me. May need some help to see if this is still a viable option to get linting on unused class selectors. Otherwise, I think we don't lint for this. The only other tools I have found are ones that remove unused class selectors, which I don't think we want to do.

**Comment by averdin2** [comment]
The GitHub action appears to work properly for forked repo pull requests on the main repo.

**Comment by Sihemgourou** [comment]
@averdin2 
Please add update
1. Progress
2. Blockers
3. Availability
4. ETA

**Comment by averdin2** [comment]
Progress: I have been done with the bulk of this work for some time now. I was looking to add a specific style rule that is not included in stylelint's rules, but I am thinking that it might just be best to not include that rule, as it could add complications to making changes to new rules in the future.

Blockers: I need to speak with Akib about the progress I have made on this issue, and we need to come up with a way to implement this into the repo's GItHub Actions.

Availability: I am now fairly available this week. I will reach out to Akib to schedule a time to discuss this issue.

ETA: As soon as I discuss this issue and my progress with Akib, I can move forward with the implementation.

**Comment by averdin2** [comment]
Progress: I have implemented the GitHub action on my own website repo and everything works well. The last thing I need to do is connect the GitHub action to Hack for LA's website repo and make a PR for the new files I made to make this work. I most likely need to consult @Aveline-art in implementing GitHub actions on the Hack for LA website repo.

Blockers: None.

Availability: I think I am going to bring this up in the July 15th meeting and see if there is time to either get this done then, otherwise I will coordinate any discussions needed.

ETA: This week

**Comment by averdin2** [comment]
### Notes
 - In the linter.yml file, on line 22 the `VALIDATE_ALL_CODEBASE: false` means that the GitHub action only checks committed files to the repo. The reason why I am not checking the entire codebase is that I do not believe it is necessary to run a check on the entire codebase for every pull request. I am going to make an issue and deal with all the errors that are in the current codebase as a result of running `VALIDATE_ALL_CODEBASE: true`. I believe after fixing these issues, only checking the files in new pull requests should be sufficient for linting the codebase.
 - The types of rules I implemented are only those whose failure would likely result in actual style errors like not enacting the proper styles. I did not add any rules that limit features or require more strict enforcement of particular code formatting rules. If linting rules like these wanted to be added in the future, we can do this by adding rules to the .stylelintrc.json file.
 - To add linting for another language (like JavaScript), we can use the same GitHub action to do so. (See [here](https://github.com/github/super-linter/blob/master/docs/using-rules-files.md))

**Comment by averdin2** [comment]
Progress: I have made a pull request for this issue #1957 , but I decided to make a separate issue to deal with all the linting issues in the codebase currently. I am also using that issue #1958 as a secondary means to test all the rules I have implemented in the GitHub action.

Blockers: None, but like #1958, this issue is technically a low priority, and I am assisting on issue #1833 which is a high priority. 

Availability: I attend most meetings and am responsive on Slack if there are any questions regarding this issue.

ETA: 2 weeks to finish up work on #1958 and make changes to rules to reopen the pull request.

**Comment by averdin2** [comment]
Progress: Made plan for review. I still need to make some changes before I can open up the PR again.
Blockers: None, I am just working on this a bit slowly.
Availability: I attend most meetings and am responsive on Slack if there are any questions regarding this issue.
ETA: I should make my changes by this weekend and then open up the PR.

**Comment by averdin2** [comment]
Update: Just waiting on a review for this PR. I also am planning on adding one more commit tomorrow.
